src/systems/scene_broadcaster/SceneBroadcaster.cc

```diff
@@ -257,8 +257,54 @@ class ignition::gazebo::systems::SceneBroadcasterPrivate
   /// \brief Flag used to indicate if periodic changes need to be published
   /// This is currently only used in playback mode.
   public: bool pubPeriodicChanges{false};
+
+  /// \brief Stores a cache of components that are changed. (This prevents dropping of
+  /// periodic change components which may not be frequent enough)
+  public: std::unordered_map<Entity,
+    std::unordered_map<ComponentTypeId,
+    std::unique_ptr<components::BaseComponent>>> changedComponents;
```

Comment:
I see we're storing clones of the component. But this will incur unnecessary clones since if there are multiple changes to the component between SceneBroadcaster publications, we'd be replacing the cloned component with new clones multiple times. I think it would be more efficient if we just  stored the `ComponentTypeId` and get the actual component based on the type id when we're ready to publish the state message.

Reply:
Thanks for pointing this out I've found a cleaner way to do this in [467048c](https://github.com/gazebosim/gz-sim/pull/2010/commits/467048c5a309ae6e1fbf3e3ab27e51d225ca8145).